ledger-tool/src/main.rs

```diff
@@ -2796,6 +2815,41 @@ fn main() {
                     }
                 });
         }
+        ("repair-roots", Some(arg_matches)) => {
+            let blockstore = open_blockstore(
+                &ledger_path,
+                AccessType::TryPrimaryThenSecondary,
+                wal_recovery_mode,
+            );
+            let start_root = if let Some(height) = arg_matches.value_of("start_root") {
+                Slot::from_str(height).expect("Before root must be a number")
+            } else {
+                blockstore.max_root()
+            };
+            let end_root = if let Some(height) = arg_matches.value_of("end_root") {
+                Slot::from_str(height).expect("Until root must be a number")
+            } else {
+                blockstore.lowest_slot()
+            };
+            assert!(start_root > end_root);
+            assert!(blockstore.is_root(start_root));
+            let ancestor_iterator =
+                AncestorIterator::new(start_root, &blockstore).take_while(|&slot| slot >= end_root);
+            let roots_to_fix: Vec<_> = ancestor_iterator
+                .filter(|slot| !blockstore.is_root(*slot))
+                .collect();
+            if !roots_to_fix.is_empty() {
+                blockstore.set_roots(&roots_to_fix).unwrap_or_else(|err| {
+                    eprintln!("Unable to set roots {:?}: {}", roots_to_fix, err);
+                    exit(1);
+                });
+            } else {
+                println!(
```

Comment:
nice. I like placing exceptional clause at `else` too even at the cost of extra `!` for the `if` condition. :)

Reply:
Hehe, well I originally wrote it with no `!`, ie. `if roots_to_fix.len() > 0`, but clippy prefers `is_empty`